### ISSUE TITLE
Hook periods/delays

### ISSUE THREAD

**Issue by JDGrimes** [WordPoints/hooks-api#10]
This is a book I accidentally wrote on the subject while taking notes:
## What are periods?

Periods/delays represent an amount of time which must pass between hook
firings. If the hook is fired more than once in the period, the target will
still be hit just once.
## What do periods relate to hooks?

Periods relate to the hook itself. The period may also be related to the
target and/or other hook argument(s). The two defining factors are thus the
unique hook identifier (which would be composed of the instance ID and hook
slug), and the identifiers for any related args.
## How are periods calculated?

Periods could be calculated in two different ways:
1. First, the period could be
   calculated absolutely in reference to the unix epoch. A 24 hour period would
   thus always start at the same time every day. We could factor timezones in so
   that the period start coincided with the start of the day.
2. The delay could also be calculated relative to the last time the hook was
   fired. In this case, a 24 hour period would begin when the hook fires. Day
   boundaries and timezones are irrelevant.

Each of these options could be useful, depending on the circumstances. For
example, the absolute periods are the behavior users expect from the Periodic
points hook, while a user might want to only award points for one comment left
by a user per-minute.

This latter case brings up several questions.
## How do periods interact with limits?

Consider a case where a user wants to limit the number of times each user is
awarded points for commenting. They don't want to set an absolute limit, but
a relative limit, that is, a limit within a given period. For example, they
might want to only award a user for up to 3 comments in a five-minute period.
## How do periods sometimes relate to args?

Consider a case where a user wants to award points to each user only once per
minute for commenting on a post, but still wants to award points for
commenting on another post within that period. In this case the period is
relative to a particular argument, the post_id.
## How are periods-related data stored?

In order to facilitate this feature the start time of the period needs to be
stored. There are several things to consider:
- We can't just store this as one of the arg object's metadata, because there
  may be 0 or several args.
- As objects get deleted and periods expire, lots of bloat could occur.

One possible approach to solving these issues would be to use transients to
store the data. That way things would be automatically cleaned up once the
periods expired. However, there is an issue here: transients are not
guaranteed to persist as long as their expiration.

We can still use a transients-like approach, but we'd need to create our own
API. Alternatively, we could save the data as hook metadata, but we'd need to
manufacture our own API for that as well, and it would still be subject to
bloat (or take a lot of work to avoid it).
## What periods-related data needs to be stored?

In addition to how these are stored, we need to think about what data is being
stored. Absolute periods and relative delays aren't exactly the same, so let's
consider them separately.

Periods have a start date and an end date. These are both known at the time
that the hook fires, and are calculated absolutely, that is, in reference to
the unix epoch 0. There are therefore three pieces of data that we could
store: the period start time, the period end time, and the hook fire time. It
is also possible that we'll want to store more information in the future, like
a limit (see above). Which of these three we store needs to take into account
several things, including what the expected behavior would be when a hook's
settings change and the period length changes. For example, if we have a hook
that has a period of 1 day, and we change it to 1 hour, what would happen?
- If we just store the period start time, the period length will always be
  calculated relative the current settings, not any prior settings. However,
  the period won't take into account the last fire time, which could lead to
  unexpected behavior in this case. If the user leaves a comment at 11:30 pm
  and receives points for it, then the settings get changed from 1 day to 1 hr,
  the period start would be 23.5 hours ago. So if the user leaves another
  comment at 11:35, he'll be awarded points for that comment too, since the
  period start was more than 1 hr ago.
- If we just store the period end, the period length will always be calculated
  relative to the previous settings. This seems to dodge the above problem, but
  it actually just inverts it. Instead, when the user leaves a comment at 1:00
  am, and receives points, and the current period length is 1 day, the period
  would end in 23 hrs. But now if the period length is changed from 1 day
  to 1 hr, the user still won't be able to earn points for comments for the
  next 23 hrs.
- If we just store the last fire time of the hook, the period will always be
  calculated relative to that (but still absolutely). This is the way that it
  was previously done, and it avoids the problems with setting changes. However
  by doing this, we must consider that the period data won't have an expiration
  date, and therefore we will not be able to clean the database based on that.

We need to either find a way to work around the issue of storage bloat, or a
way to work around the issue of unexpected behavior when settings changed if
we just store the period end time. Let us first consider solutions to the
later problem.

If we just store the expiration, we could mitigate the problem of things not
working right when the settings change by updating all of the expiration times
when the settings are changed. That way, in the scenario where a comment is
left at 1 am and the expiration is set to 23 hours later, when the settings
were updated to be just one hour, we could set the expiration to be just 1 hr
later. However, if the settings change wasn't made until 3 am, the user would
end up loosing out, because instead of being able to gain points for a comment
immediately, he'd have to wait until 4 am. So, in this case, we'd still need
to store the time that the last hook fired. But we also need to consider the
case where the settings call for a period of 1 hr. When the user leaves a
comment at 1 am the expiration would be 2 am. Then the settings are changed to
1 day at 1:30 am, so the expiration would be updated to 12 midnight. That is
all good. But consider the case where the comment is left at 1 am and so the
expiration is 2 am, but the settings aren't updated until 3 am. Now the
expiration will have passed, so the data for that expiration may have been
deleted, meaning that we can't update the expiration. That ultimately will
mean that the user is able to post a second time that day after 3 am, even
though the period is now 1 day. This is kind of an edge-case, but it could
confuse users. We might also want to consider the possibility that in the
future periods could be limits of the type 5 per period or 3 per period, etc.,
instead of just 1 per period. This can will suffer from the same issue. For
example, if a change is made from 1 per-hour to 2 per-day, and a user has
left a comment just over an hour ago, they will still be able to leave 2 more
comments that day. The only way to fix that issue however would be to keep
track of all hook firings from the beginning of time.
## Periods and Retroactivity

I guess we really need to realize that periods are by nature retroactive, and
while that is something we hope to make a part of this API, it isn't yet there
So, until it is, I propose that we store the expiration time and the time of
the last firing, and later on we can work around the bug outlined above using
the retroactivity API.
## Table structure

So, in order to build this API, we're going to need a way of storing the data.

This will be comprised of the following:
- The expiration time.
- A unique identifier based on the concerned args (possibly a hash).
- The last time the hook was fired.
- Probably other metadata in the future.

We only need to query based on two of these, the expiration and identifier. I
think we can probably make the last fire time a part of the array of metadata.
So, in order to store it, we're going to need a table with 3 fields. That
means that options aren't an option.

Actually the id probably won't be a hash string, because we might need to get
that information in a usable form if other hook settings change and we want
to take that into account retroactively. But that again will require the
retroactive API, so it is beyond what were after currently (but still good to
keep in mind).

So we need to have a field with the expiration time. Another field could be
the hook ID (this would consist of the hook slug and the instance ID, or these
could be two separate fields). Because the arg info will need to be an array,
we'll want to store them with the other metadata. Then we'll also need to
store the values of the args. This and the hook ID field are what we'd query
on to locate the hook, and they could in fact be one field. This field could
be a hash, but as it is noted above, we'd still need to have the values in the
metadata too in case they are needed for retroactive stuff (though this could
use more though, and we'll remove this from the metadata if it turns out not
to be needed). So we are down to three fields:
- Expiration
- ID hash
- Metadata

So, for this we'll have to create a new table. It may also be useful to have
an ID field so that we can identify each row.
## Multiple Periods

It's also true that more than one period could be useful per-hook. That may be
beyond what we'll incorporate into the UI at first, but I think we should
consider supporting it in the API, or at least design the API in such a way
that it can be added later if we desire.

For an example of how this might be used, consider the user_visit hook, which
fires whenever a user visits a site. It uses the current_user_id arg. On
multisite, the site_id arg is also available. This gives us two different args
which we could attach periods to. We might, for example, want each site on the
network to award points to only a single visitor per-minute. At the same time,
we might want each user to be awarded points no more than once per-hour. This
allows each site to award 60 different users per-hour, all different. In this
configuration, each user could only earn points once per-hour irrespective of
what sites they visited. However, we could change the period on the
current_user_id arg to be a compound period incorporating the site_id arg as
well. Now each user would be able to earn points once per-hour per-site. This
demonstrates how multiple periods could end up incorporating the same arg when
compound periods are used.

**Comment by JDGrimes** [comment]
## API

The current design of the table:

```
id BIGINT(20) UNSIGNED NOT NULL AUTO_INCREMENT,
hook_id VARCHAR(255) DEFAULT NULL,
arg_hash VARCHAR(255) DEFAULT NULL,
expiration BIGINT(20) UNSIGNED NOT NULL DEFAULT 0,
meta LONGTEXT,
PRIMARY KEY  (id),
KEY hook_id (hook_id(191))
```
- Allowance is made for multiple periods per reaction.
- There is currently no support for limits.
- There is no API for registering period lengths.
- I'm unconvinced that we need to store period metadata, and that if we do we should do it this way.
- We don't currently update the expirations when the settings of the period change.
- We don't store the last fire time, which means we cover less edge-cases for absolute periods and are unable to properly handle relative periods.

### Problems

#### Absolute Periods

##### Length Decrease

One problem for absolute periods, pointed out above, is that when the period length shrinks, and we update the future expirations down to the new period end, firing will not be possible until that first period has ended, even if the fire took place before it would have started. The suggestion for fixing this would be to also save the last firing time, that way we could just delete the started periods for fires that took place before the beginning of the (new, shorter) period.

But would that work properly with limits added? We would actually need the time for each fire,  so we know which fires actually occurred during the current period and need to be counted. (Same thing when limits change.) I guess we could use the meta to store that.

##### Length Increase

The other issue is when increasing the period length of absolute periods. If the expired periods are already cleaned up and no longer in the database, we can't find the past expirations and update them to the future period expiration. The result is that recent past fires that would have fallen within the current period won't be taken into account, and so a second firing within the period can take place.

##### Solutions

For the length decrease, we just need to start saving the last fire time. But to fix the length increase issues we could do several things:

###### No Overlap

One possibility I've thought of would be making it so that changes to absolute periods don't take effect until the current period ends. I'm not sure exactly how we'd handle that in the code, but it would mean the new and old period lengths wouldn't overlap, and thus the above issues would be moot. But at times the user might not want it to work that way. We could include the ability to override this, but that makes the periods more complicated for the user than they already can be.

This would also mean that when increasing the period length, from an hour to a day, for example, you'd need to wait several hours, until the end of the day, to start the next period.

At first I thought this might simplify things, but at the moment I think it would just make them more complex

###### Don't Delete Expired Periods

This could be mitigated by increasing the time between database cleanings to something like once a week, although the edge-case would still effect periods of greater lengths. But the issue is less likely to be noticed then, and might also be less severe. (And actually, I haven't any code to clean up the database yet anyway, so there isn't currently a set time period.)

Of course, if we wanted to we could go all the way and just never clean up expired periods. The problem is that this would lead to large amounts of database bloat, with very little benefit. If we're going to do that, it might be better to just ignore this edge-case, or... 

###### No Absolute Periods

Relative periods also have edge-cases, but they seem less severe. However, we need to support absolute periods, because they are used in the points hooks, and we need to maintain back-compat so that we can import those to the new API.

With that in mind, we don't really handle these edge cases in the points hooks API, so perhaps we don't have to solve them all in the initial version of the new API. But it's good to understand what they are and what the possible solutions are, at least.

**Comment by JDGrimes** [comment]
## Reversals and Spam

Unlike the conditions extension, the periods extension needs to listen for reversals and spam. When one of those takes place, we need to delete any open periods matching the event args.

We have built the API to make that possible, but the periods extension doesn't currently listen for reversals or spam. So that's something we need to implement.

## Settings Changes

When the args a period is related to change, this will immediately invalidate all started periods, meaning that we essentially start from scratch. Honestly, that seems like correct behavior, and I'm really not sure that there is much else that we can do. We could try to find the open periods and update them, but how would we know which open periods matched the updated period args? We just save the hash of the arg values that we can query by, but we don't have the arg values when updating, only at firing time. All we have when updating is the signatures of the args themselves. Even if we save the signatures of the args for the open periods in the period meta, how do we know which old signatures should correspond to which new signatures? They could be somewhat similar or very different, so we'd really just have to guess based on the order (assuming multiple different periods per reaction).

I think that for now it is safer just to let the periods reset when the settings change.

I was going to say that we could just delete the open periods when that happens, but I think that actually it might make sense to keep them around, in case the user decides to switch back to the original period settings. (We could still clean them up as we normally would when they expire, of course.)

**Comment by JDGrimes** [comment]
## UI

There is currently no UI for this, so I'm going to begin working on some of the skeleton for that. Perhaps that will also help us solve some of the UI issues.

### Name

One thing that I'm not sure about is what to actually call this in the UI (and so possibly the name of the API will be changing too). Calling these things "Periods" doesn't really convey to the user exactly what they are used for. Something like "Delays" or "Limits" seems much more self-explanatory.

The term "Limits" would be somewhat ambiguous though, since we're planning to introduce a limits API eventually, one not tied to time periods. To avoid confusion, I guess we could call these "time limits" or "rate limits". The latter sounds better, but might not convey the absolute nature of the time period calculations, for those familiar with the term "rate limiting". (I've always assumed that rate limiting uses relative time periods, and I believe that it usually does.) "Time limits" sounds too much like you're going to limit the amount of time that the reaction will work for, which isn't the case at all.

The term "Delays" also conveys rather strongly the idea that the periods will be relative, not absolute.

Adding an option for toggling absolute/relative calculation might help make it clear that these periods can be absolute (relative to the clock), but  I think that the term "delay" to encompass all of them would still be too confusing, especially if we use absolute periods by default.

For some background, the term period comes from the old Periodic points hook, which fired periodically when a user visited the site. (Instead of firing every single time they visited, it would only fire once per period.)

I think the term "rate limits" is the best so far, but it clashes with both the limit and rates APIs (see #5 and #6). Of course, I don't really like the term "rate" for the rate API at the moment either, so that will likely change.

We'd still be clashing with the limit API though. Maybe we should just combine that API with this one, making adding the time limit optional. So you could have a limit of once per user, or once per user per year.

I guess it kind of makes sense that these different extensions would work together. I'm not sure that they would really work correctly at all if they were separate. What would happen, for example, if there was a limit of twice per user, and a rate limit of once per user per day? I guess it would work fine. They would work fine together or separate, so I think it makes sense to leave them separate, instead of complicating things by putting them together. On the other hand, the limits will need to use a log database too, or the retroactive API. So there is some similarity between them.

I'm also curious now how we'll get the "rate" API to work together with these in an "obvious" way. How will we make it obvious to users that rates are applied after periods but before limits? But I guess it really wouldn't make sense for it to work any other way.

**Comment by JDGrimes** [comment]
## API: Reversals

When a reversal occurs how do we know that the original was awarded points? How do we know that the original fire occurred in the current period?

First I'll answer the latter question. If we don't find any active periods in the database, then we know that the event being reversed must not have occurred during the current period. There is nothing for us to do anyway, since what we would do is subtract one fire from the period (and if there's no period there then there's nothing to subtract from).

But that still leaves the question as to how we know that the event actually caused a fire the first time, that was counted as one of our fires for that period. What if it occurred after the limit was reached? (That applies even if we don't really have limits so that the limit is always 1.) To do that we'd need to know the time that the original event happened. The only way that we can know that is if we save a log of each event that hits during each period and a means of identifying it.  We could save this information as period metadata. To identify each event individually we'd need to have a signature from all the values associated with that event, not just for the args linked to the period. I think actually that we wouldn't need to loop through and get all the child values, just the values of the top-level args, but I'm not 100% sure.

This makes me think of another question: what if a user is able to modify one of the relationships of an entity that the periods relate to? Would that allow them to keep changing that value and re-triggering the event to get more points? Do we even stop events from multi-firing at all? Actually, no, we don't. See #53.

I'm wondering now whether it would make sense to log all fires. That would help solve the issues both in #53 and here.

**Comment by JDGrimes** [comment]
## API: Absolute Period Length Changes

Is the behavior for length increases really other than what users might expect? When the length is increased we don't go back and retroactively revert points transactions beyond the limit which had already occurred during the period, so why should we count them toward the current period at all? Period changes just aren't retroactive. So it sort of makes sense to just wait until the next period starts to apply the changes, except that the user expects the changes to apply immediately. A better move might be to divide the period length by the limit for that period, and compare that to the time left in the period. So if the period is a month and the limit is 15 and there are 4 days left, each use would be allowed 4 / 30 \* 15 = 2 / 15 \* 15 = 2 hits before the end of the period. This makes a lot of sense.

Let's simplify things by leaving out these limits. The limit is always once per period then. So if we change from once per day to once per month, should we just allow one more hit no matter where we are in the month? But then users who have already gotten their hit for that day would be allowed a second hit the same day, unless we add special handling for that. Or maybe we'd just check if a user had received any hits in that month at all, if they had, they won't be able to receive anymore. If they hadn't, they'd still be able to receive one. That sounds like what we should do.

So with limits added instead of allowing every user just two more hits for the month, we'd maintain a cap of 15 hits for the whole month. So a user that already had 20 hits that month wouldn't be able to get any more, a user with 5 hits would be able to get 10 more, a user with 0 would be allowed up to 15 more, etc. In some ways that doesn't seem fair for the users who already hit their cap, because the users who didn't get to fill their quota in a very short time (a single hour or less, even). But on the other hand, they've already got more than they'd normally be allowed anyway, so I guess its not really unfair. It's just what happens when you change the rules in the middle of the game.

But what about length _decreases_? Are we handling them in a retroactive manner? We need to be consistent here. Yes, length decreases work correctly, all they do is make the period end sooner. The only issue is when we also change the limit (assuming that we allow this).

When the limit is modified we'd just do like I said above with the increases, check whether we've already passed the limit in the current period, and allow up to that amount if we haven't. There will not be retroactive reverting of transactions.

**Comment by JDGrimes** [comment]
## API: Which Args?

There are several questions about the API in regards to which args periods may relate to.

### Identities Only?

Should periods only be allowed to relate to entity identities, or all arg types? 

For example, should we be able to set a period on a reaction to the post publish event related to the post title, or just the post itself? I guess it really doesn't make sense to set the period on the post title, since that is something that is usually unique to each post anyway. But I suppose that the user might want to prevent multiple awards per-period for the same title. It would make much more sense with some other attribute, of course. Any attribute which has a specific set of possible values might indeed be useful in relating to periods. But I think it seems silly to offer all possible entities, as we do with conditions, when only a very few will actually be useful. So perhaps we should limit it to identities by default, and then consider how to expand that as needed in the future. Perhaps the best solution is just to register entities for such fields, so that they would have identities.

### Entity Arrays

Assuming that we limit periods to identities, that still leaves open another question: how do we handle entities that are in one-to-many relationships?

For example, we might want to relate the period to the post tags. So then we could have a period that would only award points once per week for post tag. So only one post with that tag would be awarded per week. But would we calculate that based on all of the post tags collectively, or each tag individually?

Taking all of the post tags into account would mean that if one post had the "ice" and "snow" tags and another had the "snow" and "snowflake" tags, both posts would be awarded. If we ran the periods per individual tag, we'd award only the first post, because they both have the "snow" tag.

I guess these really are both possible, because we could offer both the  `Post » Tags` and `Post » Tags » Tag` options.

All of this would take special handling of course, so I'm not sure that we want to pursue it at the moment. I see no reason why we can't punt it until a later version, and wait until we get all the kinks out of the rest of the periods API first.

### Summary

For now we'll pursue identities only, with no support for entity arrays. But we'll be keeping this in mind and may well add support for both of these later, though likely not in the initial version of the API.

**Comment by JDGrimes** [comment]
## API: Absolute Periods

Absolute periods seem strange. I mean, no matter how long the period is, it is always possible for a timely user to gain double the hits in just a few seconds. All they have to do is trigger the reaction just before and just after the end of the period. Yes, they still can only get the set number of hits per period, but they don't have to put in double effort to do that. It seems like rate limiting should work differently than that. It's hard to believe that this is what users would expect.

I guess the real issue here is that there are two different uses for periods. One is as a rate limiter, as a means of handling spam. Another is to just limit the number of times you give out points, for whatever reason.

For the later, absolute periods work well, for the former delays/throttling works better.

We could offer both options, but that complicates things for the user, forcing them to make a decision that they may not fully understand. Rather than forcing them to do that, it would be better if we could make that decision for them, saving their brainpower for more important things.

We can do that in several ways:
- Always defaulting to one or the other (absolute or relative).
- Defaulting to one sometimes and the other at other times, based on the circumstance.
- Coming up with some sort of blend that isn't absolute, and isn't relative, but in between.

The most simple of these is the first, and that alone is a big plus. However, my concern would be that users find that it doesn't work as they'd expect in some situations.

Automatically switching to throttling mode in certain situations doesn't sound like a bad idea, the only problem is that it adds complexity that could catch a user off guard. Also, we'd have to decide what situations to do that in. Any period under 5 minutes? Only for specific events?

The final suggestion above would be to do something like this: calculate the periods relative to the clock, but add a throttling period at the start that amounts to some particular fraction of the period length, say, 10% (possibly we'd also take the limit into account, if we do limits). So if there was a hit near the end of the last period, we'd have to wait a bit before counting a hit in this period. For example, if the period length was an hour and the hit came in the last minute, we'd have to wait about 6 minutes (10% of 60 minutes) into the new hour before counting another hit. If the hit in the previous period came 4 minutes before the end of the hour, we'd have to wait about 2 minutes into this hour, etc.

All of this is especially considering the case when periods are tied to the reaction target. In other cases it really isn't so important, since a single user cannot game the system to get double points (unless they use multiple accounts and can donate points back and forth). Actually, I guess they could, but there's nothing stopping another user from getting a hit in one of the periods instead. In fact it is quite likely that the first period will already have reached its limit, so the user would probably try to come in right at the start of the next period to get the awarded hit.

For some purposes it might make more sense to take a random draw from the events that occurred in the period to decide who to award (this would let a user game the system by triggering lots of events, although that could be mitigated by just giving each target only a single chance in the lottery).

Once again this brings us to exactly what we're trying to accomplish with this. What is the purpose of periods. Are they a means of combating spam? Are they just a fun way to limit the number of points handed out during a given time span?

To answer this question, I guess we should look back at the Periodic points hook. It implicitly relates the period only to the target user. Much simpler than the huge number of possibilities we've been imagining. There wasn't really any concept of "spam" to it. It was just a way to encourage users to spend time on the site regularly. It also provided a way to give them a regular top-up of points. The use-cases that we have been imagining are largely different. They would fall into the category of discouraging something, rather than encouraging it (that's what throttling is all about). This puts us in a strange position, since we're encouraging something by giving points for doing it, while at the same time being very limited in that encouragement. This is really just asking for users to game the system, to engage in a race for who is first. If you're site is supposed to be a game, that's fine, and may actually be a really nice feature. But otherwise it seems to me that it could easily become a nightmare.

Just because we can imagine all of these different uses for periods, doesn't mean that they would actually be that useful. It also doesn't mean that periods need to cover all of those use-cases; perhaps there are better solutions to some of them. We can decide that some of those use-cases, while technically possible, are actually wrong, and not what the API is designed to do. Periods are currently too complex. They need to do some basic things, and do them well. Then we can see what users actually want, and decide whether periods should do that, or whether a new API is needed.

At the moment I am tempted to make periods an event-specific thing. Only events that ask for periods to be used would have them enabled in the UI. So we'd enable them for the User Visit event, but not for other events.

I'm also thinking that we should limit periods to relating only to the target of the reaction, nothing else.

And if we do those things, I think we should just make the periods absolute for now, for back-compat if for no other reason.

Also, we'd obviously only be allowing for one period per reaction in the UI, since the period would always be related to the target.

**Comment by JDGrimes** [comment]
## UI: Name

I've been contemplating the name a bit more, and the only other word that might fit would be "interval", but I don't think it's any better than "period". I'm going to go with "rate limit" in the UI for now. We won't change the API yet though.

**Comment by JDGrimes** [comment]
## API: Clashing with Limits

I mentioned the possibility of this API clashing with the limits API above. If we only implement the most basic scope, we won't have any trouble with that, but I did want to note this observation in case we come back to that later:

A limit of 1 per user doesn't make sense when combined with a period of 2 per user per hour. However, it's not going to cause anything to break. Also, it could make perfect sense when combined with a different period, like 2 per site per hour.

**Comment by JDGrimes** [comment]
For posterity: when I was using the term "last fire" above, I was talking about the last hit, not just any fire.

**Comment by JDGrimes** [comment]
## API: Period clean-up

I'm currently considering not automatically cleaning up periods. The reason is because having old periods is useful when handling settings changes. However, I don't see why we'd ever have a period longer than a year, so deleting periods from more than a year ago should always be OK. Just in case longer periods are registered, we could get the list of registered periods lengths and just delete periods older than the longest one. That wouldn't help us when registering a new period of greater length, however, since if we change the settings of a reaction we won't have periods older than that to check against.

By "older than x" above, I mean that the hit time is longer ago than that, not the expiration. (Although I suppose that if we introduce limits we'd need to consider the latest hit time, not the original hit time.) I think actually that we could just store the hit times, and forget about the expirations. Because when we store the expirations we need to update them when the settings change, whereas when storing hit times we don't need to update anything.

The only reason we would want to store the expirations would be for knowing when it was safe to delete a period, but like I said above, we can calculate that based on the hit time just as easily. For absolute periods the hit time wouldn't always need to be saved for the full length of the period, and could actually be deleted when a new period begins. But for relative periods we'd need to save them for the full length, so it might make sense just to save them for the full time of the longest length and then delete them, if we clean them up at all.

The benefit of clean up is that it keeps the database from bloating. This provides multiple benefits. Obviously will less bloat there is less data being stored, which alone is a good thing. Having less data in the database can also make queries faster, which is important to us since we'll likely be running these queries on every single page-load. 

However, I suppose that the table will always have less data in it than the points logs table, so it isn't like it will become more huge than anything else the plugin uses.

One idea that I had, which was kind of hinted at above when I mentioned the possibility of deleting absolute periods when the period ends: instead of deleting periods, we could leave them there but just update the hit times and reuse them. This means that we'd only ever have one period per event per signature (which is the target by default). We wouldn't have older periods, but we wouldn't need them anyway unless we introduce limits. Right now the limit is 1, so we only need to know if _any_ hits occurred within the period, not how many. So we currently only need to keep 1 period per event per target/signature around.

I think that for now, it is safest if we just don't worry about cleaning up the periods. It is something that we can revisit later as this API becomes more resolved. If we're only going to delete those more than a year old anyway, we have a year before the code would be doing anything. :-)

**Comment by JDGrimes** [comment]
> This makes me think of another question: what if a user is able to modify one of the relationships of an entity that the periods relate to? Would that allow them to keep changing that value and re-triggering the event to get more points?

The answer is yes, although this is not important when we're only relating periods to the target. It is something that we will need to consider if we expand this API, however.

**Comment by JDGrimes** [comment]
## API: Reversals

Above we asked these two questions, but did not answer them very well:

> When a reversal occurs how do we know that the original was awarded points? How do we know that the original fire occurred in the current period?

I'll answer the first question first. When an event fires, it will either hit or miss the target. If it hit the target then a period will be created. If it misses no period will exist. Thus the existence of a matching period would seem to indicate that the event created that period and therefore the period should be deleted.

The issue is that the event being reversed may have occurred multiple times. The first occurrence would have caused the period to be added to the database, while the later occurrences would be ignored. But we should only remove the period from the database once all of those fires have been reversed.

By "fires", I mean events that would have hit the target except for being blocked by the period. In other words, these will be misses, but we wouldn't count all misses, because the other misses wouldn't have hit the target even if the period didn't exist; that is, we're only interested in misses which would have been hits if the period hadn't already started. Unfortunately, the API doesn't really provide us an easy means of identifying these.

I think the only way to really handle this properly would be to log hits as in #55. Then each hit would be assigned a unique ID, which would be passed to the `after_hit()` method. When a reversal occurred, we'd then pass the IDs of the hits being reversed to the reversal methods. (I'm still not sure exactly how we'd match the reversals to particular hits, but that's a bridge we can burn when we come to it. I guess we'd always assume that it was the latest matching hit that was being reversed, based on the signature of the main event args.)

The point is that right now we really don't have the framework to properly handle reversals. At the same time, we don't actually _need_ to handle reversals, since the User Visit event cannot be reversed by default. It is really something that we could punt until later. I'm not entirely convinced that's the best thing to do, but I'm not sure we can do much with this in the initial version.

**Comment by JDGrimes** [comment]
Reversals make the data structure more complex. When doing a reversal, we don't know the reaction ID or the args a period is related to. All we have is the event slug and all of the event args. If we want to do reversals we need to add the event slug and event signature (calculated from the top-level args of the event) to the database table, so we can query by those. But once we have a list of events, we still don't know whether they actually match, until we calculate the period signature from the related args. To know the related args we'd have to save that information as well (perhaps in meta).

There are also other added complexities. Like the fact that if we used the args for an event as its signature, when a new arg is registered for an event, all existing periods for that event would no longer respond to reversals. (The same would occur if we removed an arg from an event.) The only way to solve that would be to save the signature as a raw array instead of a hash, and then check only that all values in both arrays matched, ignoring the keys that were absent in one of the arrays. That makes things very difficult, since we need to introduce a new table or we'll be unable to run simple queries and would instead have to pull all of the data out and check each period for that event slug.

All in all, that's more than we need to mess with when we're not even using this feature. Not supporting reversals doesn't mean that the extension can't be used with reversible events, only that when those events are reversed/spammed the user might have points removed and won't be able to gain them back during the period. It could be frustrating to users, but at least it doesn't open up a way to game the system.

In short, trying to provide the base groundwork for handling periods is not going to give us much benefit, and won't be at all "basic". We'd do better to wait until we actually need the feature and can better assess the potential pitfalls and how to handle them.

**Comment by JDGrimes** [comment]
## Summary

In summary, here's what we did:
- No limits (the limit is always 1).
- Periods relate to the reaction target by default.
- In the UI, the periods extension is only enabled for specific events.
- The ability to add multiple periods per-reaction is not exposed in the UI.
- So in the UI the periods automatically relate to the reaction target.
- And the ability to relate a period to multiple args is also not exposed in the UI.
- Periods are absolute by default. This is not exposed in the UI.
- We store the time of the hit (not the expiration as originally proposed).
- And we do not delete old periods.
- This means that when the length of the period increases, older hits are taken into account.
- When the length decreases, hits that are older are ignored.
- The period signature is calculated based on the values of the arg(s) to which the period relates.
- The periods are not related to event slugs specifically, so the behavior when the event a reaction relates to changes is undefined.
- Reversals are not listened for, so behavior for non-repeatable events is currently undefined.

We might expand upon this in the future, but probably not before the initial release of the API.

## TODO
- [x] There are no tests for caching.
- [x] I think we could probably benefit from adding an index to the database table on the signature and reaction, since we query by that.
- [x] We need to add a filter or other means of adding custom period lengths.
- [x] We need to allow the "Award each user no more than once per" string be supplied per-reactor.

Anything beyond that will probably need to be in a new ticket.

**Comment by JDGrimes** [comment]
The current UI (this is the bottom of the settings for a User Visit reaction):

![wordpoints_ _points_types_ _wp_trunk_ _wordpress_-_2015-12-23_17 05 00](https://cloud.githubusercontent.com/assets/4005415/11985907/964d6520-a997-11e5-9e83-a990d3bdafe8.png)

**Comment by JDGrimes** [comment]
While working on #56, I realized that we need to be able to differentiate between reactions from different reactors, since IDs are not unique across reactors. We need to add a `reactor` column to the table, or else the reaction IDs from different reactors will be confused.

If we implement #55 of course, we'd just reference the hit ID and let all of this information be stored in the hits table.

**Comment by JDGrimes** [comment]
> - The periods are not related to event slugs specifically, so the behavior when the event a reaction relates to changes is undefined.

This might change if we do #55, depending on whether the event slug is taken into account when past hits are queried.

**Comment by JDGrimes** [comment]
> > - The periods are not related to event slugs specifically, so the behavior when the event a reaction relates to changes is undefined.
> 
> This might change if we do #55, depending on whether the event slug is taken into account when past hits are queried.

Indeed, this is exactly what has happened.
